src/nvs.rs

```diff
@@ -736,6 +806,89 @@ impl RawHandle for EspNvs<NvsEncrypted> {
         self.1
     }
 }
+impl RawHandle for EspNvs<NvsDefault> {
+    type Handle = nvs_handle_t;
+
+    fn handle(&self) -> Self::Handle {
+        self.1
+    }
+}
+
+#[cfg(all(
+    esp_idf_version_major = "5",
+    not(any(esp_idf_version_minor = "1", esp_idf_version_minor = "0"))
+))]
+pub struct EspNvsKeys<'a> {
+    // The EspNvs must not be dropped while the iterator is still in use,
+    // this reference ensures that.
+    _nvs: PhantomData<&'a ()>,
+    raw_iter: nvs_iterator_t,
+    is_exhausted: bool,
+}
+
+#[cfg(all(
+    esp_idf_version_major = "5",
+    not(any(esp_idf_version_minor = "1", esp_idf_version_minor = "0"))
+))]
+impl<'a> Iterator for EspNvsKeys<'a> {
+    type Item = (heapless::String<16>, Option<NvsDataType>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.is_exhausted || self.raw_iter.is_null() {
+            return None;
+        }
+
+        let mut info: nvs_entry_info_t = Default::default();
+        match unsafe { nvs_entry_info(self.raw_iter, &mut info as *mut _) } {
+            ESP_ERR_NVS_NOT_FOUND => {
+                self.is_exhausted = true;
+                None
+            }
+            ESP_OK => {
+                // For the next iteration, the iterator must be advanced to the next entry,
+                // otherwise it will return the same entry again.
+                //
+                // This function call will fail if the iterator is
+                // - null, which is checked before this call
+                // - exhausted (if it is, it will set self.raw_iter to null and iteration will stop)
+                //
+                // For convenience, the error is ignored here, because it should never happen anyway.
+                // The usage example in C simply stops the iteration on error too and does not do any
+                // error handling.
+                let _ = esp!(unsafe { nvs_entry_next(&mut self.raw_iter as *mut _) });
+
+                Some((
+                    from_cstr(&info.key[..]).try_into().unwrap(),
+                    NvsDataType::from_nvs_type(info.type_),
+                ))
+            }
+            // The nvs_entry_info only fails if any of the arguments are null.
+            // The nvs_entry_info is never null, and self.raw_iter is checked for null before the invocation.
+            //
+            // Therefore this should never happen.
+            err => unimplemented!(
```

Comment:
since we think it never happens, maybe we should use unreachable here, instead of unimplemented?